From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 22 Apr 2026 02:17:13 +0800
Subject: [PATCH] SparklyPaper: Skip "distanceToSqr" call in
 "ServerEntity#sendChanges" if the delta movement hasn't changed

Co-authored by: MrPowerGamerBR <git@mrpowergamerbr.com>
As part of: SparklyPaper (https://github.com/SparklyPower/SparklyPaper/blob/a69d3690472c9967772ffd4d4bd1f41403b7ea6f/sparklypaper-server/minecraft-patches/features/0002-Skip-distanceToSqr-call-in-ServerEntity-sendChanges-.patch)

diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 17fdc01dae1d13b2d6f886e6e2a9cdeb310a0923..f01624b3932aad4ed51042dc3b6ac216590a64d2 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -199,6 +199,7 @@ public class ServerEntity {
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) {
                     Vec3 movement = this.entity.getDeltaMovement();
+                    if (movement != this.lastSentMovement) { // SparklyPaper start - skip distanceToSqr call in ServerEntity#sendChanges if the delta movement hasn't changed
                     double diff = movement.distanceToSqr(this.lastSentMovement);
                     if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
                         this.lastSentMovement = movement;
@@ -216,6 +217,7 @@ public class ServerEntity {
                             this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                         }
                     }
+                    } // SparklyPaper end
                 }
 
                 if (packet != null) {
